src/components/EditorEmbedModal/EditorEmbedModal.tsx

```diff
@@ -0,0 +1,181 @@
+import {
+  FormControl,
+  HStack,
+  Modal,
+  ModalBody,
+  ModalContent,
+  ModalFooter,
+  ModalHeader,
+  ModalOverlay,
+  Text,
+  Textarea,
+} from "@chakra-ui/react"
+import { yupResolver } from "@hookform/resolvers/yup"
+import {
+  Button,
+  FormErrorMessage,
+  FormLabel,
+  Link,
+  ModalCloseButton,
+} from "@opengovsg/design-system-react"
+import * as Cheerio from "cheerio"
+import { useEffect } from "react"
+import { FormProvider, useForm } from "react-hook-form"
+import * as Yup from "yup"
+
+import { useCspHook } from "hooks/settingsHooks"
+
+import { isEmbedCodeValid } from "utils/allowedHTML"
+
+import { EditorEmbedContents } from "types/editPage"
+
+interface EditorEmbedModalProps {
+  isOpen: boolean
+  onClose: () => void
+  onProceed: (embedCode: EditorEmbedContents) => void
+  cursorValue: string
+}
+
+export const EditorEmbedModal = ({
+  isOpen,
+  onClose,
+  onProceed,
+  cursorValue,
+}: EditorEmbedModalProps): JSX.Element => {
+  const { data: csp } = useCspHook()
+
+  const handleSubmit = (embedCode: EditorEmbedContents) => {
+    const { value } = embedCode
+    const $ = Cheerio.load(value)
+
+    if ($("blockquote").hasClass("instagram-media")) {
+      const postUrl = $(".instagram-media").attr("data-instgrm-permalink")
+      const url = document.createElement("a")
+      url.href = postUrl || ""
+      const code = `<iframe src="https://${url.host}${url.pathname}embed/" frameborder="0" allowfullscreen="true" width="320" height="440"></iframe>`
+      onProceed({ value: code })
+    } else {
+      onProceed(embedCode)
+    }
+  }
+
+  const methods = useForm<EditorEmbedContents>({
+    mode: "onTouched",
+    resolver: yupResolver(
+      Yup.object().shape({
+        value: Yup.string()
+          .required(
+            "Content to embed cannot be empty. Please enter valid HTML code."
+          )
+          .matches(
+            // Blockquote is to allow for Instagram embeds
+            /^(<iframe |<blockquote )(.*)$/s,
+            "Please enter a valid embed code"
+          )
+          .test({
+            name: "isEmbedCodeValid",
+            message:
+              "We detected unauthorised JavaScript in the code. Please remove the JavaScript content and try again.",
+            test: (value) => isEmbedCodeValid(csp, value),
+          }),
+      })
+    ),
+    defaultValues: {
+      value: "",
+    },
+  })
+
+  useEffect(() => {
+    if (isOpen) {
+      methods.setValue(
+        "value",
+        cursorValue
+          .replace('<div class="iframe-wrapper">', "")
+          .replace("</div>", "")
+      )
+    } else {
+      methods.reset()
+    }
+  }, [cursorValue, isOpen, methods])
+
+  return (
+    <Modal isOpen={isOpen} onClose={onClose}>
```

Comment:
previously we didnt have a modal for embed right? is this because we want to capture the user input and validate it before it reaches the editor?

Reply:
Yup, that's the first reason. The second reason is that users are not able to edit the embed code after it has been inserted into the editor. With this modal, they can now do that.